### PR TITLE
KAFKA-17108: Add EarliestPendingUpload offset spec in ListOffsets API

### PR DESCRIPTION
**Pull request by abhijeetk88**
This is the first part of the implementation of
[KIP-1023](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1023%3A+Follower+fetch+from+tiered+offset)

The purpose of this pull request is for the broker to start returning
the correct offset when it receives a -6 as a timestamp in a ListOffsets
API request.

Added unit tests for the new timestamp.

Reviewers: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

### CODE REVIEW

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 1]
```diff
@@ -1334,6 +1334,37 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
+                val epochOpt = leaderEpochCache.get.epochForOffset(earliestPendingUploadOffset)
+                if (epochOpt.isPresent) {
+                  Optional.of(epochOpt.getAsInt)
+                } else {
+                  Optional.empty()
+                }
+              } else {
+                Optional.empty()
+              }
+            Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, earliestPendingUploadOffset, epochResult))
```
Comment (kamalcph) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]:
```suggestion
            val epochResult = leaderEpochCache.flatMap(_.epochForOffset(earliestPendingUploadOffset).asScala)
              .asJava.asInstanceOf[Optional[Integer]]
            Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, earliestPendingUploadOffset, epochResult))
```

File: core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala [hunk 2]
```diff
@@ -2222,6 +2222,273 @@ class UnifiedLogTest {
...
+      timestamp = secondTimestamp),
+      leaderEpoch = leaderEpoch)
+
+    assertEquals(Some(new TimestampAndOffset(ListOffsetsResponse.UNKNOWN_TIMESTAMP, -1, Optional.of(-1))),
+      log.fetchOffsetByTimestamp(ListOffsetsRequest.EARLIEST_PENDING_UPLOAD_OFFSET_TIMESTAMP))
+  }
+
+  @Test
+  def testFetchEarliestPendingUploadTimestampWithRemoteStorage(): Unit = {
+    val remoteLogManager = mock(classOf[RemoteLogManager])
```
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala]: we can refactor the test to extract the "prepare" phase which is duplicated across multiple tests.

**Review by kamalcph — COMMENTED** [review]
Thanks for the patch! Left minor comments to address.

Should we need to add an OffsetSpec on the client side to make the `earliest_pending_upload_offset` queryable from command line tools similar to 6e324487fa93c540ef611eb9c9694c464fdecd3c?

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 1 above)
Comment (abhijeetk88) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: Done. Thanks for the suggestion.

File: core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala (hunk 2 above)
Comment (abhijeetk88) [reply core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala]: Done

**Comment by abhijeetk88** [comment]
> Should we need to add an OffsetSpec on the client side to make the `earliest_pending_upload_offset` queryable from command line tools similar to [6e32448](https://github.com/apache/kafka/commit/6e324487fa93c540ef611eb9c9694c464fdecd3c)?

@kamalcph I can make it as part of a separate change. Will that be ok?

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 3]
```diff
@@ -1334,6 +1334,27 @@ class UnifiedLog(@volatile var logStartOffset: Long,
         } else {
           Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, -1L, Optional.of(-1)))
         }
+      } else if (targetTimestamp == ListOffsetsRequest.EARLIEST_PENDING_UPLOAD_OFFSET_TIMESTAMP) {
```
Comment (OmniaGM) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: Can we extract this into different method and call it from here especially that the logic here has too many if stataments ? I know the rest of fetchOffsetByTimestamp has been done in-place but the method has grown and it hard to follow.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 1 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: There are other places in the same method can benefit from this refactor as well like  line 1319 and 1298

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 4]
```diff
@@ -49,6 +49,8 @@ public class ListOffsetsRequest extends AbstractRequest {
 
     public static final long LATEST_TIERED_TIMESTAMP = -5L;
 
+    public static final long EARLIEST_PENDING_UPLOAD_OFFSET_TIMESTAMP = -6L;
```
Comment (kamalcph) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: consider renaming it to `EARLIEST_PENDING_UPLOAD_TIMESTAMP` for uniformity with other names.

**Review by kamalcph — COMMENTED** [review]
Overall LGTM. Left one minor comment to address. Thanks for covering the patch with detailed unit tests!

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 3 above)
Comment (abhijeetk88) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: Done

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java (hunk 4 above)
Comment (abhijeetk88) [reply clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Done

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 1 above)
Comment (abhijeetk88) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: Made the requested changes. Also changed for LATEST_TIMESTAMP

**Comment by abhijeetk88** [comment]
@kamalcph @OmniaGM addressed your comments. PTAL again.

**Review by kamalcph — COMMENTED** [review]
LGTM, thanks for addressing the review comments!

Should we have to bump the ListOffsetRequest version for this change? PTAL.

cc @showuon @satishd

**Comment by kamalcph** [comment]
@abhijeetk88

Please rebase this PR with trunk.

**Comment by abhijeetk88** [comment]
I would like to revive this PR. @kamalcph @showuon will you be able to help with reviews. I can rebase the PR with trunk.

File: core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala [hunk 5]
```diff
@@ -2416,6 +2416,193 @@ class UnifiedLogTest {
...
+    // Test initial state before any records
+    assertFetchOffsetBySpecialTimestamp(log, None, new TimestampAndOffset(ListOffsetsResponse.UNKNOWN_TIMESTAMP, -1, Optional.of(-1)),
+      ListOffsetsRequest.EARLIEST_PENDING_UPLOAD_TIMESTAMP)
+
+    // Append records
+    val _ = prepareLogWithSequentialRecords(log, recordCount = 2)
+
+    // Test state after records are appended
+    assertFetchOffsetBySpecialTimestamp(log, None, new TimestampAndOffset(ListOffsetsResponse.UNKNOWN_OFFSET, -1, Optional.of(-1)),
+      ListOffsetsRequest.EARLIEST_PENDING_UPLOAD_TIMESTAMP)
```
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala:2433]: should this be updated to? 

ListOffsetsResponse.UNKNOWN_OFFSET -> ListOffsetsResponse.UNKNOWN_TIMESTAMP?

**Review by kamalcph — COMMENTED** [review]
Thanks for the patch! Please address the below comments:

1. ListOffsetsRequest version needs a bump as a new targetTimestamp value is added . Previous [comment](https://github.com/apache/kafka/pull/15213#discussion_r1552178574) and
2. Also, could you add the EarliestPendingUploadSpec in [OffsetSpec](https://sourcegraph.com/github.com/apache/kafka/-/blob/clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java?L30) and cover with integration tests?

**Comment by kamalcph** [comment]
@abhijeetk88 

Could you follow this https://github.com/apache/kafka/pull/16781 PR and handle similarly for EarliestPendingUploadSpec? Thanks!

**Review by kamalcph — APPROVED** [review]
LGTM, thanks for the patch!

**Comment by chia7712** [comment]
@abhijeetk88 Do you mind documenting these new APIs in `upgrade.html`?

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 6]
```diff
@@ -91,4 +92,13 @@ public static OffsetSpec earliestLocal() {
     public static OffsetSpec latestTiered() {
         return new LatestTieredSpec();
     }
+
+    /**
+     * Used to retrieve the earliest offset of records that are pending upload to remote storage.
+     * <br/>
+     * Note: When tiered storage is not enabled, we will return unknown offset.
+     */
+    public static OffsetSpec earliestPendingUpload() {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:101]: I mean this new public API

Comment (abhijeetk88) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:101]: sure, will add.
